gdal/frmts/grib/gribdataset.cpp

```diff
@@ -2287,15 +2309,34 @@ void GRIBDataset::SetGribMetaData(grib_MetaData *meta)
         if (rPixelSizeY < 0 || fabs(rPixelSizeY - meta->gds.Dy) > 0.002)
             rPixelSizeY = meta->gds.Dy;
 
-        // Longitude origin of GRIB files is sometimes funny. Try to shift as close
-        // as possible to the traditional [-180,180] longitude range
-        // See https://trac.osgeo.org/gdal/ticket/7103
-        if( ((rMinX >= 179 && rPixelSizeX * meta->gds.Nx > 10) || rMinX >= 180) &&
+        // GRIB2 files have longitudes in the [0-360] range
+        // Shift them to the traditional [-180,180] longitude range
+        // See https://github.com/OSGeo/gdal/issues/4524
+        if (meta->GribVersion == 2 &&
+            (rMinX + rPixelSizeX >= 180 || rMaxX >= 180) && 
             CPLTestBool(CPLGetConfigOption("GRIB_ADJUST_LONGITUDE_RANGE", "YES")) )
         {
-            CPLDebug("GRIB", "Adjusting longitude origin from %f to %f",
-                     rMinX - rPixelSizeX / 2, rMinX - rPixelSizeX / 2 - 360 );
-            rMinX -= 360;
+            if (rPixelSizeX * nRasterXSize > 360)
+                CPLError(CE_Warning, CPLE_NotSupported,
+                    "Cannot properly handle GRIB2 files with overlaps and 0-360 longitudes");
+            else if (rPixelSizeX * nRasterXSize == 360 && meta->gds.projType == GS3_LATLON)
+            {
+                nSplitAndSwap = static_cast<int>(ceil(180 / rPixelSizeX));
```

Comment:
I don't think that formula is correct in the general case. It should probably be
```suggestion
                nSplitAndSwap = static_cast<int>(ceil((180 - rMinX) / rPixelSizeX));
```
and should be conditionalized to 180 >= rMinX

Reply:
It is conditionalized (these are nested ifs) and in this special case I set rMinX = -180 hardcoded + the half pixel adjustment a few lines below
This is in fact the special case (which is in fact the most common case) of a GRIBv2 that covers the whole world - this one I translate to -180 to 180

But if I am to make antimeridian-spanning GeoTIFFs, this needs generalization

Reply:
> It is conditionalized (these are nested ifs)

I may miss something, but it seems to be that if we have for example rMinx = 90 and rMaxX = 90 + 360, we would go here, hence we cannot assume that rMinX = 0. Or if we want to limit to that case, it should be a condition of that branch

Reply:
I believe my above point is still valid and not addressed

Reply:
Unless I am missing something, when this statement is evaluated, those hold true:
```
rMinX == 0
rMinX + rPixelSizeX >= 180 || rMaxX - rPixelSizeX >= 180
```

Reply:
can you point where rMinX == 0 is checked ?

Reply:
I check(ed) for `rPixelSizeX * nRasterXSize == 360` which (normally) would mean `rMinX == 0` and then I was setting it to -180 hardcoded value
But indeed, maybe there could have been an (invalid) GRIB going from lets says 2 to 362 that was not going to work with that - but that GRIB would have been broken anyway